components/render/sdl2backend.cpp

```diff
@@ -818,203 +820,117 @@ namespace Render
-    }
-    
-    // returns: whether or not to keep drawing the current line (see where it's used for what I mean by "line"
-    bool drawLevelHelper(const Level::Level& level, SpriteGroup& minSprites, int32_t x, int32_t y, int32_t levelX, int32_t levelY)
-    {
-        // if the top left of our screen is on a negative coord (we're at the edge of the map)
-        // we want to continue the line, but we obviously don't want to draw tiles outside the map
-        if(x < 0)
-            return true;
-        
-        if((size_t)x < level.width() && (size_t)y < level.height())
+           auto point = startingPoint;
+           auto tile = startingTile;
+
+           while (point.x < WIDTH + tileWidth / 2)
+             {
+               point.x += tileWidth;
+               ++tile.x;
+               --tile.y;
+               processTile (tile, point);
+             }
+        };
+
+      // then from top left to top-bottom
+      while (startingPoint.y < HEIGHT + staticObjectHeight - tileHeight)
         {
-            size_t index = level[x][y].index();
-            int32_t xCoord = (y*(-32)) + 32*x + level.height()*32-32 + levelX;
-            int32_t yCoord = (y*16) + 16*x + levelY;
-
-            if(xCoord <=  WIDTH && yCoord <= HEIGHT)
-            {
-                if(index < minSprites.size())
-                    drawAt(minSprites[index], xCoord, yCoord);
-                
-                return true;
-            }
-            
-            return false;
+          ++startingTile.y;
+          startingPoint.x -= tileWidth / 2; startingPoint.y += tileHeight / 2;
+          processLine ();
+          ++startingTile.x;
+          startingPoint.x += tileWidth / 2; startingPoint.y += tileHeight / 2;
+          processLine ();
         }
-        
-        return false;
     }
 
-    void drawLevel(const Level::Level& level, size_t minTopsHandle, size_t minBottomsHandle, SpriteCacheBase* cache, LevelObjects& objs, int32_t x1, int32_t y1, int32_t x2, int32_t y2, size_t dist)
-    {
-        int32_t levelX, levelY;
-        getMapScreenCoords(level, x1, y1, x2, y2, dist, levelX, levelY);
+    void drawLevel(const Level::Level& level, size_t minTopsHandle, size_t minBottomsHandle, SpriteCacheBase* cache, LevelObjects& objs, int32_t x1, int32_t y1, int32_t x2, int32_t y2, size_t dist) {
+      auto toScreen = worldToScreenVector({x1, y1}, {x2, y2}, dist);
+      SpriteGroup* minBottoms = cache->get(minBottomsHandle);
+      auto isInvalidTile = [&](const Tile &tile){ return tile.x < 0 || tile.y < 0 || tile.x >= static_cast<int32_t> (level.width()) || tile.y >= static_cast<int32_t> (level.height());};
 
-        //TODO clean up the magic numbers here, and elsewhere in this file
+      // drawing on the ground objects
+      drawObjectsByTiles (toScreen, [&](const Tile &tile, const Point &topLeft){
+          // Fill invalid tiles with ground, it looks ok but it's probably better to have something else than zero-eth sprite here
```

Comment:
Eventually I'd like to rig up infinite outward spread of some tileset-appropriate value, but for now this is fine.